### PR TITLE
(#1185) Add hook scripts

### LINKED ISSUE #1185: Allow post/pre hooks to run on upgrade/install/uninstall
The idea for this have been taken from pacman (the main package manager for arch linux).
One usage example on how it is used in pacman would be running reflector after pacman-mirrorlist package is updated: https://wiki.archlinux.org/index.php/Reflector#Pacman_Hook

What I request is similiar: have a directory (either configurable or hard coded to `$env:ChocolateyInstall\hook` that stores a bunch of powershell scripts that are meant to be run after/before a package update/install (or after choco have completed the full update/install of all packages).

The way I'm thinking this could work is through a naming convention. ie: `(pre|post){pacakge-id}-upgrade/install/uninstall.ps1` or `(pre|post)global-upgrade/install/uninstall.ps1`.

The reason I would like to see something like this implemented is because quite a few times I have a need/want to automatically configure some extra configuration for the package(s) that I'm installing (usually user specific configurations) (could range from anything as easy as removing a shortcut, to full configuration of git, virtualbox, smartgit, or even choco. as well as others).

### PR DESCRIPTION
**Pull request by TheCakeIsNaOH**
## Description Of Changes

This adds the ability to run Powershell scripts (hook scripts) before
and after each type of package automation script runs. The hook scripts
can be for all packages, or for a specific package ID. They are looked
for recursively inside `$env:ChocolateyInstall\hooks`

The filename of the `.ps1` file determines when they are run:
`<pre|post>-<install|beforemodify|uninstall>-<packageID|all>.ps1`

The first section is for timing of when the hook script runs.
Ore scripts run before the package automation script
(e.g. the ChocolateyInstall.ps1), while post scripts are run afterwards.
The second section of the filename determines what type of automation
script the hook is supposed to run for. The third section either
a string of the package ID the hook is to run for, or  the keyword all
to specify that it should be run for all package IDs.

If a hook needs to be run for more than one of these cases and/or it is
complex, the logic can be extracted out to another file or module and
dotsourced/imported.

If there are hook script(s) found, they will be run even if there is no
package automation script.

The hooks are run in the same context as package automation scripts
are, so they have access to the same environment variables as the
package scripts do. 

It also adds the ability for choco to install `.hook` packages that contain
hook PowerShell scripts, the same in concept as a `.template` or `.extension`
package.

For packages ending in `.hook`, any files or directories inside a top
level folder called "hook" will be copied to a directory under
`$env:chocolatey\hooks\` named the same as the package ID with the
`.hook` stripped off the end (e.g `remove-shortcuts` for a package ID of `remove-shortcuts.hook`).

During upgrades, this directory will be removed, and updated files will
be copied over. During uninstalls, this directory will be removed.

Finally, there is a command line option added to skip hook scripts. 

## Motivation and Context

See #1185 for most of it

This does not add hooks before pack/push, because validation is planned to be added directly into Chocolatey, https://github.com/chocolatey/choco/issues/602 

Hooks that run pre/post everything (e.g. not for each individual package) are not added here, but they could be added in the future more easily with the infrastructure added here.

## Testing

Most cases covered by added unit and integration tests.

Manual testing:
- Pack and install the `scriptpackage.hook` from the integration test context
- Run `choco install wget` and ensure the pre and post install hooks run
- Uninstall wget 
- Run `choco install wget --skip-hooks`, and ensure that the hooks do not run, but the `chocolateyInstall.ps1` does.
- Uninstall wget 
- Run `choco install wget --skip-powershell`, and ensure that neither the hooks nor the `chocolateyInstall.ps1`  runs.

## Change Types Made

* [ ] Bug fix (non-breaking change)
* [x] Feature / Enhancement (non-breaking change)
* [ ] Breaking change (fix or feature that could cause existing functionality to change)

## Related Issue

Fixes #1185

## Change Checklist

* [x] Requires a change to the documentation
* [ ] Documentation has been updated - will need to add docs, TODO to open an issue on docs repo
* [x] Tests to cover my changes, have been added
* [x] All new and existing tests passed.
* [x] PowerShell v2 compatibility checked.

### CODE REVIEW

**Comment by TheCakeIsNaOH** [comment]
This could use another set of eyes on it to ensure I've not broken anything with regards to the `chocolateyScriptRunner.ps1`, and to double check that it is PowerShell v2 compatible. 

I also could use suggestions for any other automated tests that should be added.

**Review by corbob — COMMENTED** [review]
I've only gotten to starting looking at the `chocolateyScriptRunner.ps1` file so far, but we will also want to ensure any packages that are updated or added here also get added to our internal repository after this is merged.

As for PowerShell 2 support, that should be tested as part of the Pester tests, although it's possible that we don't fully and/or properly test the `chocolateyScriptRunner.ps1` file with those tests. I'll try to get a Windows 7 VM out to test against this.

File: src/chocolatey.resources/helpers/chocolateyScriptRunner.ps1
Comment (TheCakeIsNaOH) [reply src/chocolatey.resources/helpers/chocolateyScriptRunner.ps1:18]: @vexx32 thoughts on this?

**Comment by TheCakeIsNaOH** [comment]
> but we will also want to ensure any packages that are updated or added here also get added to our internal repository after this is merged.

I'm not quite sure what you mean by this? 

If it is the packages added to the integration testing context, those are used only for integration testing, and are build/used only as part of that testing.

**Comment by corbob** [comment]
> > but we will also want to ensure any packages that are updated or added here also get added to our internal repository after this is merged.
> 
> I'm not quite sure what you mean by this?
> 
> If it is the packages added to the integration testing context, those are used only for integration testing, and are build/used only as part of that testing.

What I mean is that currently all of the packages are built automatically in the vagrant environment for any Pester tests that might use them. When we run them within Test Kitchen, they operate against a repository that expects packages to be there. For consistency sake, the package will need to be built by us and pushed to that testing repository in case we ever try to use it from the Pester tests.

**Review by corbob — CHANGES_REQUESTED** [review]
I haven't yet gotten to the tests yet. Just a few comments and suggestions (and at least one note for me to come back to for my own testing even though the more I think about it the more I think it does exactly what we expect it to do 🤔 )

reviewing the tests will be a tomorrow task.

File: src/chocolatey.resources/helpers/chocolateyScriptRunner.ps1
Comment (corbob) [reply src/chocolatey.resources/helpers/chocolateyScriptRunner.ps1:18]: Yeah, I've spent too much time in Pester land with HashTables and objects that don't print well without `Out-String`. This should be just fine, my basic test script (below) output the same on 2 and 5.1.

```powershell
$arr = @(1, 2, 3, 'hello', 'goodbye', 4, 5)
Write-Host $arr
# Output: 1 2 3 hello goodbye 4 5
```

File: src/chocolatey.tests.integration/scenarios/UninstallScenarios.cs
Comment (corbob) [inline src/chocolatey.tests.integration/scenarios/UninstallScenarios.cs:1072]:
```suggestion
            private PackageResult _packageResult;

```
Should be consistent with private variable naming.

File: src/chocolatey.tests.integration/scenarios/UninstallScenarios.cs
Comment (corbob) [inline src/chocolatey.tests.integration/scenarios/UninstallScenarios.cs:1151]:
```suggestion
            private PackageResult _packageResult;

```
Consistent variable naming.

**Review by corbob — CHANGES_REQUESTED** [review]
Tests overall look good. We'll want to add some Pester tests to cover these things as well.

Only things that stood out was the variable naming in a few of the tests. Might be good to check over all of them to ensure I didn't miss any, and obviously: do the rename in Visual Studio and ensure it renames correctly.

File: src/chocolatey/infrastructure.app/services/PowershellService.cs
Comment (TheCakeIsNaOH) [reply src/chocolatey/infrastructure.app/services/PowershellService.cs]: Yup, that make sense, renamed.

File: src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs
Comment (TheCakeIsNaOH) [reply src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs:1541]: No reason, I've made them consistent.

File: src/chocolatey/infrastructure.app/services/PowershellService.cs
Comment (TheCakeIsNaOH) [reply src/chocolatey/infrastructure.app/services/PowershellService.cs]: Fixed.

File: src/chocolatey.tests.integration/scenarios/InstallScenarios.cs
Comment (TheCakeIsNaOH) [reply src/chocolatey.tests.integration/scenarios/InstallScenarios.cs]: I agree that they should be consistent, so I've renamed them to use the underscore. 
However, in the integration tests, there was a pre-existing mix of `packageResult` and `_packageResult`, so that should be fixed at some point.

**Review by corbob — APPROVED** [review]
LGTM! I'll get this in line with develop and then get it merged.

**Review by corbob — APPROVED** [review]

**Comment by corbob** [comment]
Thanks for getting this fixed up @TheCakeIsNaOH!

**Pull request merged**
